File: android/app/src/main/kotlin/com/example/panel_app/LoginPayload.kt
package com.example.panel_app

import android.util.Base64
import org.json.JSONArray
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest
import java.util.zip.InflaterInputStream
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Port of contoh/ninja_sage/login_payload.py to Kotlin.
 *
 * This module rebuilds SystemLogin.loginUser parameters from
 * username/password, character_seed and character_key.
 */

data class LoaderInfo(
    val bytesLoaded: Int = 8_216_461,
    val bytesTotal: Int = 8_216_461,
)

object LoginPayload {

    private const val DEFAULT_LIBRARY_URL =
        "https://ns-assets.ninjasage.id/static/lib/library.bin"

    private fun readAll(input: InputStream): ByteArray {
        val buffer = ByteArray(8 * 1024)
        val out = ByteArrayOutputStream()
        var read: Int
        while (true) {
            read = input.read(buffer)
            if (read <= 0) break
            out.write(buffer, 0, read)
        }
        return out.toByteArray()
    }

    private class Crypt {
        companion object {
            private fun makeIv(seed: String): ByteArray {
                val iv = seed.toByteArray(Charsets.ISO_8859_1)
                val blockSize = 16
                val padLen = blockSize - (iv.size % blockSize)
                val padded = ByteArray(iv.size + padLen)
                System.arraycopy(iv, 0, padded, 0, iv.size)
                for (i in iv.size until padded.size) {
                    padded[i] = padLen.toByte()
                }
                return padded.copyOf(blockSize)
            }

            fun encrypt(
                plaintext: String,
                key: String,
                seed: Int,
            ): String {
                val aesKey = key.toByteArray(Charsets.ISO_8859_1)
                val iv = makeIv(seed.toString())
                val secretKey = SecretKeySpec(aesKey, "AES")
                val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
                cipher.init(
                    Cipher.ENCRYPT_MODE,
                    secretKey,
                    IvParameterSpec(iv),
                )
                val data = plaintext.toByteArray(Charsets.ISO_8859_1)
                val ciphertext = cipher.doFinal(data)
                return Base64.encodeToString(ciphertext, Base64.NO_WRAP)
            }
        }
    }

    private class PMPrng(seed: Int) {
        private val mod = 2_147_483_647
        private val mul = 16_807
        private var value: Int

        init {
            var s = seed
            if (s == 0) {
                val t = System.currentTimeMillis()
                val r = (Math.random() * 0.025 * 0x7FFFFFFF).toInt()
                s = ((t.toLong() xor r.toLong()) and 0x7FFFFFFF).toInt()
            }
            value = s and 0x7FFFFFFF
        }

        fun nextInt(): Int {
            value = ((value.toLong() * mul) % mod).toInt()
            return value
        }
    }

    @Volatile
    private var cachedLevels: Map<String, Int>? = null

    private fun loadLibraryLevels(
        libraryUrl: String = DEFAULT_LIBRARY_URL,
    ): Map<String, Int> {
        cachedLevels?.let { return it }

        val conn = URL(libraryUrl).openConnection() as HttpURLConnection
        conn.requestMethod = "GET"
        conn.connectTimeout = 20000
        conn.readTimeout = 20000
        val compressed = conn.inputStream.use { readAll(it) }

        val inflater = InflaterInputStream(ByteArrayInputStream(compressed))
        val decompressed = inflater.use { readAll(it) }
        val text = String(decompressed, Charsets.UTF_8)

        val items = JSONArray(text)
        val levels = mutableMapOf<String, Int>()
        for (i in 0 until items.length()) {
            val obj = items.getJSONObject(i)
            val id = obj.optString("id", null) ?: continue
            val level = obj.optInt("level", 0)
            levels[id] = level
        }

        cachedLevels = levels
        return levels
    }

    private fun cucsgHash(value: String): String {
        val payload = ByteArray(value.length) { index ->
            (value[index].code and 0xFF).toByte()
        }
        val digest = MessageDigest.getInstance("SHA-256").digest(payload)
        val sb = StringBuilder()
        for (b in digest) {
            sb.append(String.format("%02x", b))
        }
        return sb.toString()
    }

    private fun safeMod(a: Int, b: Int): Int {
        return if (b == 0) 0 else a % b
    }

    private fun getSpecificItem(
        loader: LoaderInfo,
        characterSeed: Int,
        levels: Map<String, Int>,
    ): String {
        val bytesTotal = loader.bytesTotal
        val bytesLoaded = loader.bytesLoaded
        val lvlHair1 = levels["hair_10000_1"] ?: 0
        val lvlHair0 = levels["hair_10000_0"] ?: 0
        val lvlAcc2003 = levels["accessory_2003"] ?: 0

        val loc4Num =
            (bytesTotal xor bytesLoaded) +
                1337 xor characterSeed xor 1337 +
                1337 xor characterSeed xor 1337 +
                1337 xor 0x0539 and
                safeMod(safeMod(bytesLoaded, lvlHair1), bytesLoaded) and
                bytesTotal xor
                safeMod(safeMod(lvlHair0, characterSeed), 1_333_777) +
                lvlAcc2003

        val loc4Str = loc4Num.toString()
        val hashed = cucsgHash(loc4Str)
        val seedStr = characterSeed.toString()
        return seedStr + hashed + seedStr.repeat(4)
    }

    private fun getRandomNSeed(
        characterSeed: Int,
        loader: LoaderInfo,
    ): String {
        val seedRng = characterSeed % loader.bytesLoaded
        val rng = PMPrng(seedRng)
        return buildString {
            repeat(4) {
                append(rng.nextInt())
            }
        }
    }

    fun buildLoginComponents(
        username: String,
        password: String,
        characterSeedRaw: Number,
        characterKey: String,
        loader: LoaderInfo = LoaderInfo(),
        libraryUrl: String = DEFAULT_LIBRARY_URL,
    ): Map<String, Any> {
        val characterSeed = characterSeedRaw.toInt()
        val levels = loadLibraryLevels(libraryUrl)
        val encryptedPassword =
            Crypt.encrypt(password, characterKey, characterSeed)
        val specificItem = getSpecificItem(loader, characterSeed, levels)
        val randomSeed = getRandomNSeed(characterSeed, loader)

        return linkedMapOf(
            "username" to username,
            "encrypted_password" to encryptedPassword,
            "character_seed" to characterSeed,
            "bytes_loaded" to loader.bytesLoaded,
            "bytes_total" to loader.bytesTotal,
            "character_key" to characterKey,
            "specific_item" to specificItem,
            "random_seed" to randomSeed,
            "password_length" to password.length,
        )
    }
}

File: contoh/ninja_sage/login_payload.py
"""Helpers to rebuild SystemLogin.loginUser payloads from username/password."""

from __future__ import annotations

import base64
import hashlib
import json
import random
import time
import urllib.request
import zlib
from dataclasses import dataclass
from functools import lru_cache
from typing import Dict

from Crypto.Cipher import AES
from Crypto.Util.Padding import pad

DEFAULT_LIBRARY_URL = "https://ns-assets.ninjasage.id/static/lib/library.bin"


@dataclass(slots=True)
class LoaderInfo:
    """Mimics ``loaderInfo`` from the Flash client."""

    bytes_loaded: int = 8_216_461
    bytes_total: int = 8_216_461


class _Crypt:
    @staticmethod
    def _make_iv(seed: str) -> bytes:
        iv = seed.encode("latin1")
        block_size = 16
        pad_len = block_size - (len(iv) % block_size)
        padded = iv + bytes([pad_len]) * pad_len
        return padded[:block_size]

    @classmethod
    def encrypt(cls, plaintext: str, key: str, seed: int) -> str:
        aes_key = key.encode("latin1")
        iv = cls._make_iv(str(seed))
        data = pad(plaintext.encode("latin1"), 16)
        cipher = AES.new(aes_key, AES.MODE_CBC, iv)
        ciphertext = cipher.encrypt(data)
        return base64.b64encode(ciphertext).decode("ascii")


class _PM_PRNG:
    MOD = 2_147_483_647
    MUL = 16_807

    def __init__(self, seed: int) -> None:
        if seed == 0:
            t = int(time.time() * 1000)
            r = int(random.random() * 0.025 * 0x7FFFFFFF)
            seed = (t ^ r) & 0x7FFFFFFF
        self.seed = seed & 0x7FFFFFFF

    def next_int(self) -> int:
        self.seed = (self.seed * self.MUL) % self.MOD
        return self.seed


@lru_cache(maxsize=2)
def load_library_levels(library_url: str = DEFAULT_LIBRARY_URL) -> Dict[str, int]:
    with urllib.request.urlopen(library_url) as resp:
        compressed = resp.read()
    data = zlib.decompress(compressed)
    items = json.loads(data.decode("utf-8"))
    levels: Dict[str, int] = {}
    for entry in items:
        item_id = entry.get("id")
        if item_id is None:
            continue
        levels[item_id] = int(entry.get("level", 0))
    return levels


def _cucsg_hash(value: str) -> str:
    payload = bytes((ord(ch) & 0xFF) for ch in value)
    return hashlib.sha256(payload).hexdigest()


def _safe_mod(a: int, b: int) -> int:
    return 0 if b == 0 else a % b


def get_specific_item(loader: LoaderInfo, character_seed: int | float, levels: Dict[str, int]) -> str:
    bytes_total = int(loader.bytes_total)
    bytes_loaded = int(loader.bytes_loaded)
    character_seed = int(character_seed)
    lvl_hair_1 = levels.get("hair_10000_1", 0)
    lvl_hair_0 = levels.get("hair_10000_0", 0)
    lvl_acc_2003 = levels.get("accessory_2003", 0)

    loc4_num = (
        (bytes_total ^ bytes_loaded)
        + 1337
        ^ character_seed
        ^ 1337
        + 1337
        ^ character_seed
        ^ 1337
        + 1337
        ^ 0x0539
        & _safe_mod(_safe_mod(bytes_loaded, lvl_hair_1), bytes_loaded)
        & bytes_total
        ^ _safe_mod(_safe_mod(lvl_hair_0, character_seed), 1_333_777)
        + lvl_acc_2003
    )

    loc4_str = str(loc4_num)
    hashed = _cucsg_hash(loc4_str)
    seed_str = str(character_seed)
    return seed_str + hashed + seed_str * 4


def get_random_n_seed(character_seed: int | float, loader: LoaderInfo) -> str:
    character_seed = int(character_seed)
    seed_rng = character_seed % int(loader.bytes_loaded)
    rng = _PM_PRNG(seed_rng)
    return "".join(str(rng.next_int()) for _ in range(4))


def build_login_components(
    username: str,
    password: str,
    character_seed: int | float,
    character_key: str,
    *,
    loader: LoaderInfo | None = None,
    library_url: str = DEFAULT_LIBRARY_URL,
) -> Dict[str, object]:
    loader = loader or LoaderInfo()
    character_seed = int(character_seed)
    levels = load_library_levels(library_url)
    encrypted_password = _Crypt.encrypt(password, character_key, character_seed)
    specific_item = get_specific_item(loader, character_seed, levels)
    random_seed = get_random_n_seed(character_seed, loader)

    return {
        "username": username,
        "encrypted_password": encrypted_password,
        "character_seed": character_seed,
        "bytes_loaded": loader.bytes_loaded,
        "bytes_total": loader.bytes_total,
        "character_key": character_key,
        "specific_item": specific_item,
        "random_seed": random_seed,
        "password_length": len(password),
    }
